### PR TITLE
SPLICE-2379 SparkExplain Statement

### PR DESCRIPTION
**Pull request by msirek**
Add a SPARKEXPLAIN statement, similar to EXPLAIN, which prints the query execution plan, but in the case of SPARKEXPLAIN the Splice operations which are executed natively in Spark are replaced with text describing the Spark execution plan.  For example:
`sparkexplain select SUM(a1) from t1 inner join t2 on b1=b2;`

```
Native Spark Execution Plan
----------------------------
-> NativeSparkDataSet
     *HashAggregate(keys=[], functions=[sum(cast(c1#144 as bigint))], output=[c0#179L])
     +- Exchange SinglePartition
        +- *HashAggregate(keys=[], functions=[partial_sum(cast(c1#144 as bigint))], output=[sum#183L])
           +- *Project [c0#94 AS c1#144]
              +- *BroadcastHashJoin [c1#95], [c0#105], Inner, BuildRight
                 :- *Filter isnotnull(c1#95)
                 :  +- Scan ExistingRDD[c0#94,c1#95]
                       -> TableScan[T1(8064)](RS=0,totalCost=4.04, scannedRows=20,outputRows=20, outputHeapSize=60 B,partitions=1)
                 +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint)))
                    +- *Filter isnotnull(c0#105)
                       +- Scan ExistingRDD[c0#105]
                          -> TableScan[T2(8080)](RS=2,totalCost=4.04, scannedRows=20,outputRows=20, outputHeapSize=108 B,partitions=1)
```
Use the sqlshell.sh command "maximumdisplaywidth 0;" for best formatting.
Splice operations start with "->", except the "-> NativeSparkDataSet" tag, which is just an indicator that a subtree of spark operations lies below it.

Note that display of Spark operations underneath a Splice Subquery node is currently unsupported, but an abbreviated SubqueryNode subtree with one or two Splice operation child nodes is displayed as a place marker.  Compare the plan with normal EXPLAIN output to fill in the blanks.

There are 3 extra variations of SparkExplain which display different representations of the operation tree corresponding to different phases of planning and optimization in the Catalyst optimizer:

sparkexplain_logical
sparkexplain_analyzed
sparkexplain_optimized

Use these in the same manner as the regular sparkexplain statement, as in the example above.

For more details and example output, see [SPLICE-2379](https://splice.atlassian.net/browse/SPLICE-2379?oldIssueView=true).

### CODE REVIEW

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/iapi/sql/compile/CompilerContext.java [hunk 1]
```diff
@@ -656,9 +656,9 @@
 	 */
 	boolean isReferenced(SequenceDescriptor sd);
 
-    void setDataSetProcessorType(DataSetProcessorType type);
+	void setDataSetProcessorType(DataSetProcessorType type, boolean setDSPTypeinLCC);
```
Comment (dgomezferro) [inline db-engine/src/main/java/com/splicemachine/db/iapi/sql/compile/CompilerContext.java]: Why is this change needed?

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java [hunk 2]
```diff
@@ -102,6 +101,14 @@ else if (storedAs.equals("O"))
                 // The predicates have variant qualifiers (or we are reading ORC with our own reader), we couldn't push them down to the scan, process them here
                 return locatedRows.filter(new TableScanPredicateFunction<>(operationContext));
             }
+            if (dsp.isSparkExplain()) {
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java]: You could return an empty dataset instead if this could be a problem, sampling might actually read the data from Splice anyway

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java [hunk 3]
```diff
@@ -113,7 +120,19 @@ else if (storedAs.equals("O"))
         if (oneSplitPerRegion) {
             conf.set(MRConstants.ONE_SPLIT_PER_REGION, "true");
         }
-        if (useSample) {
+        if (dsp.isSparkExplain()) {
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java]: I'd either remove these blocks or make them return an empty dataset

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/RowCountOperation.java [hunk 4]
```diff
@@ -199,8 +202,31 @@ public void writeExternal(ObjectOutput out) throws IOException {
...
         long offset = getTotalOffset();
         OperationContext operationContext = dsp.createOperationContext(this);
-        DataSet<ExecRow> sourceSet = source.getDataSet(dsp).map(new CloneFunction<>(operationContext));
-        return sourceSet.zipWithIndex(operationContext).mapPartitions(new OffsetFunction<SpliceOperation, ExecRow>(operationContext, offset, fetchLimit));
+        dsp.incrementOpDepth();
+        DataSet<ExecRow> sourceDS = source.getDataSet(dsp);
+        dsp.decrementOpDepth();
+        DataSet<ExecRow> sourceSet = sourceDS.map(new CloneFunction<>(operationContext));
+        if (dsp.isSparkExplain()) {
+                        DataSet<ExecRow> ds = dsp.createDataSet(Iterators.transform(Collections.emptyIterator(),
```
Comment (dgomezferro) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/RowCountOperation.java]: Couldn't you use here dsp.getEmpty() ?

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java (hunk 2 above)
Comment (msirek) [reply hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java]: Thanks, I'll keep this in mind.  So far it's not an issue.  I'm not sure if this would change what's printed in the spark explain, for example, would it still indicate which parquet table is being read from for external tables, etc.

Comment (dgomezferro) [reply hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java]: That's a good point, it wouldn't.

File: db-engine/src/main/java/com/splicemachine/db/iapi/sql/compile/CompilerContext.java (hunk 1 above)
Comment (msirek) [reply db-engine/src/main/java/com/splicemachine/db/iapi/sql/compile/CompilerContext.java]: I think this is an old change I made to avoid using control when we're running in the Olap Server.  But I have since fixed that in CostChoosingDataSetProcessorFactory, so I will remove the above change.

Comment (msirek) [reply db-engine/src/main/java/com/splicemachine/db/iapi/sql/compile/CompilerContext.java]: I tried backing out this change, but it broke some Spark Explain tests. CostChoosingDataSetProcessorFactory.chooseProcessor() pulls the DataSetProcessorType from the lcc to determine whether to run on control or spark.  The lcc is not properly updated in all flows, so that is the intent of this change, to make sure that the DataSetProcessorType we've picked in the CompilerContext is propagated to the lcc.  Spark Explain needs to be executed on the Olap Server, so the lcc must always have the proper DataSetProcessorType setting.

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java (hunk 3 above)
Comment (msirek) [reply hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkScanSetBuilder.java]: OK, I've reverted the changes.

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/RowCountOperation.java (hunk 4 above)
Comment (msirek) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/RowCountOperation.java]: Made the change.

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by msirek** [comment]
jenkins please test branch @cdh5.13.3

**Comment by spliceci** [comment]
TEST FAILED

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1
